server/src/main/java/org/elasticsearch/snapshots/InFlightShardSnapshotStates.java

```diff
@@ -0,0 +1,140 @@
+                                            SnapshotsInProgress.ShardSnapshotStatus shardState, int shardId, String indexName) {
+        if (shardState.isActive()) {
+            busyIds.computeIfAbsent(indexName, k -> new HashSet<>()).add(shardId);
+            assert assertGenerationConsistency(generations, indexName, shardId, shardState.generation());
+        } else if (shardState.state() == SnapshotsInProgress.ShardState.SUCCESS) {
+            assert busyIds.getOrDefault(indexName, Collections.emptySet()).contains(shardId) == false :
+                "Can't have a successful operation queued after an in-progress operation";
+            generations.computeIfAbsent(indexName, k -> new HashMap<>()).put(shardId, shardState.generation());
+        }
+    }
+
+    /**
+     * Map that maps index name to a nested map of shard id to most recent successful shard generation for that
+     * shard id.
+     */
+    private final Map<String, Map<Integer, String>> generations;
+
+    /**
+     * Map of index name to a set of shard ids that currently are actively executing an operation on the repository.
+     */
+    private final Map<String, Set<Integer>> busyIds;
+
+
+    private InFlightShardSnapshotStates(Map<String, Map<Integer, String>> generations, Map<String, Set<Integer>> busyIds) {
+        this.generations = generations;
+        this.busyIds = busyIds;
+    }
+
+    private static boolean assertGenerationConsistency(Map<String, Map<Integer, String>> generations, String indexName,
+                                                       int shardId, String activeGeneration) {
+        final String bestGeneration = generations.getOrDefault(indexName, Collections.emptyMap()).get(shardId);
+        assert bestGeneration == null || activeGeneration.equals(bestGeneration);
+        return true;
+    }
+
+    /**
+     * Check if a given shard currently has an actively executing shard operation.
+     *
+     * @param indexName name of the shard's index
+     * @param shardId   shard id of the shard
+     * @return true if shard has an actively executing shard operation
+     */
+    boolean isBusy(String indexName, int shardId) {
+        return busyIds.getOrDefault(indexName, Collections.emptySet()).contains(shardId);
+    }
+
+    /**
+     * Determine the current generation for a shard by first checking if any in-flight but successful new shard
+     * snapshots or clones have set a relevant generation and then falling back to {@link ShardGenerations#getShardGen}
+     * if not.
+     *
+     * @param indexId          index id of the shard
+     * @param shardId          shard id of the shard
+     * @param shardGenerations current shard generations in the repository data
+     * @return most recent shard generation for the given shard
+     */
+    @Nullable
+    String bestGeneration(IndexId indexId, int shardId, ShardGenerations shardGenerations) {
```

Comment:
I find this method name a bit confusing. Maybe we should rename it to `generationForShard` or something along those lines?